Mathlib/Tactic/NormNum/NatFactorial.lean

```diff
@@ -0,0 +1,120 @@
+/-
+Copyright (c) 2023 Sebastian Zimmer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sebastian Zimmer
+-/
+import Mathlib.Data.Nat.Factorial.Basic
+import Mathlib.Tactic.NormNum
+
+/-! # `norm_num` extensions for factorials
+
+Extensions for `norm_num` that compute `Nat.factorial`, `Nat.ascFactorial` and `Nat.descFactorial`.
+
+This is done by reducing each of these to `ascFactorial`, which is computed using a divide and
+conquer strategy that improves performance and avoids exceeding the recursion depth.
+
+-/
+
+namespace Mathlib.Meta.NormNum
+
+open Nat Qq Lean Elab.Tactic Qq Meta
+
+lemma ascFactorial_mul_ascFactorial (n l k : ℕ) :
+    n.ascFactorial l * (n + l).ascFactorial k = n.ascFactorial (l + k) := by
+  apply mul_left_cancel₀ (factorial_pos n).ne.symm
+  simp [factorial_mul_ascFactorial, ← mul_assoc, add_assoc]
+
+lemma asc_factorial_aux (n l m a b : ℕ) (h₁ : n.ascFactorial l = a)
+    (h₂ : (n + l).ascFactorial m = b) : n.ascFactorial (l + m) = a * b := by
+  rw [← h₁, ← h₂]
+  symm
+  apply ascFactorial_mul_ascFactorial
+
+/-- Calculate `n.ascFactorial l` and return this value along with a proof of the result. -/
+partial def proveAscFactorial (n l : ℕ) : (result : Q(ℕ)) × Q(($n).ascFactorial $l = $result) :=
+  if l ≤ 50 then
+    let res : Q(ℕ) := mkRawNatLit (n.ascFactorial l)
+    ⟨res, (q(Eq.refl $res) : Expr)⟩
+  else
+    let m : ℕ := l / 2
+    let r : ℕ := l - m
+    let ⟨a, a_prf⟩ := proveAscFactorial n m
+    let ⟨b, b_prf⟩ := proveAscFactorial (n + m) r
+    let prf : Expr := q(asc_factorial_aux $n $m $r $a $b $a_prf $b_prf)
```

Comment:
`: Expr` is a bit of an antipattern when working with `Qq`; you should add a hypothesis along the lines of
```lean
have ret : Q(ℕ) := mkRawNatLit (a.natLit! * b.natLit!)
haveI : $ret =Q $a * $b := ⟨⟩
```